File: src/main/java/ru/citeck/ecos/predicate/model/NotPredicate.java
package ru.citeck.ecos.predicate.model;

import com.fasterxml.jackson.annotation.JsonProperty;

import java.util.Collections;
import java.util.List;
import java.util.Objects;

public class NotPredicate implements Predicate {

    private static final String TYPE = "not";

    @JsonProperty("val")
    private Predicate predicate;

    public NotPredicate(Predicate predicate) {
        this.predicate = predicate;
    }

    public NotPredicate() {
    }

    public Predicate getPredicate() {
        return predicate;
    }

    public void setPredicate(Predicate predicate) {
        this.predicate = predicate;
    }

    @JsonProperty("t")
    String getType() {
        return TYPE;
    }

    public static List<String> getTypes() {
        return Collections.singletonList(TYPE);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        NotPredicate that = (NotPredicate) o;
        return Objects.equals(predicate, that.predicate);
    }

    @Override
    public int hashCode() {
        return Objects.hash(predicate);
    }

    @Override
    public String toString() {
        return "NOT " + getPredicate();
    }
}

File: ecos-records/src/main/java/ru/citeck/ecos/records2/predicate/model/NotPredicate.kt
package ru.citeck.ecos.records2.predicate.model

import com.fasterxml.jackson.annotation.JsonProperty

class NotPredicate : Predicate {

    companion object {

        @JsonProperty("t")
        const val TYPE = "not"

        @JvmStatic
        fun getTypes(): List<String> {
            return listOf(TYPE)
        }
    }

    @JsonProperty("val")
    private var predicate: Predicate = VoidPredicate.INSTANCE

    constructor()

    constructor(predicate: Predicate?) {
        this.predicate = predicate ?: VoidPredicate.INSTANCE
    }

    @JsonProperty("t")
    fun getType(): String {
        return TYPE
    }

    fun getPredicate(): Predicate {
        return predicate
    }

    fun setPredicate(predicate: Predicate) {
        this.predicate = predicate
    }

    @JsonProperty("v")
    fun setVal(predicate: Predicate) {
        setPredicate(predicate)
    }

    override fun <T : Predicate> copy(): T {
        @Suppress("UNCHECKED_CAST")
        return NotPredicate(predicate.copy()) as T
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }
        if (other == null || javaClass != other.javaClass) {
            return false
        }
        val that = other as NotPredicate
        return predicate == that.predicate
    }

    override fun hashCode(): Int {
        return predicate.hashCode()
    }

    override fun toString(): String {
        return "{\"t\":\"not\",\"val\":$predicate}"
    }
}
